### DISCUSSION TITLE
Stream Offsets and Policy Max Age / Max Length

### DISCUSSION THREAD

**Discussion by thaoula** [rabbitmq/rabbitmq-server#8886] [Questions] [answered]
Hi All,

We have recently started using Streams via AMQP.  We are running on NodeJS cannot use the stream protocol.

- We have enabled a policy on the stream to prune it using Max Age and Max Length. 
- We are also persisting the x-stream-offset when we receive a new message.

The x-stream-offset appears to be integer similar to array index which is great if the steam is immutable and is never purged. 

Our issue that we are unable to resume on restart if the stream is purged due to Max Age or Max Length. The offsets are greater than the messages in the stream.

My questions are -
1. Is there a way to detect the scenario where the start offset is great than the stream x-stream-offset so we can fall back to next or first.
2. 1. If not, is there a way to obtain a stream's offset in code so that we can decided whether to start at the offset or use next or first?

Regards,
Tarek

**Comment by kjnilsson** [comment]
"The offsets are greater than the messages in the stream."

I am not sure what you mean by this?

Stream retention will always leave at least one segment worth of data - if
you try to attach at a stream offset that is smaller than the first offset
available in the stream it will fall back to the first offset already and
you don't need to do anything else.

If you have further issues please provide reproduction steps.

Cheers
Karl

On Mon, 17 Jul 2023 at 11:22, Tarek Haoula ***@***.***> wrote:

> Hi All,
>
> We have recently started using Streams via AMQP. We are running on NodeJS
> cannot use the stream protocol.
>
>    - We have enabled a policy on the stream to prune it using Max Age and
>    Max Length.
>    - We are also persisting the x-stream-offset when we receive a new
>    message.
>
> The x-stream-offset appears to be integer similar to array index which is
> great if the steam is immutable and is never purged.
>
> Our issue that we are unable to resume on restart if the stream is purged
> due to Max Age or Max Length. The offsets are greater than the messages in
> the stream.
>
> My questions are -
>
>    1. Is there a way to detect the scenario where the start offset is
>    great than the stream x-stream-offset so we can fall back to next or first.
>    2.
>       1. If not, is there a way to obtain a stream's offset in code so
>       that we can decided whether to start at the offset or use next or first?
>
> Regards,
> Tarek
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/8886>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFCUJKV4BPRMJG5WXP3XQUG6BANCNFSM6AAAAAA2MXGJMY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
*Karl Nilsson*

**Comment by thaoula** [comment]
Hi Karl, 

Thanks for your reply.

I have attached a screenshot that I hope will make a little more sense. 

Here we have a stream with only 6 items.

We also have a single consumer which has requested x-stream-offset of 101 (artificially increased here as i don't know how to force a stream to trim its messages to simulate this situation)

FYI - Our consumers, retrieve the last offset from the database and set it as an argument x-stream-offset when connecting to the stream.

The consumer will not receive a message until the stream reaches that offset and then it continues as normal. 

![image](https://github.com/rabbitmq/rabbitmq-server/assets/349133/34bd7e45-f2f5-496c-8863-6f728e1eda41)

I hope that is clearer.

Regards,
Tarek

**Comment by kjnilsson** [comment, marked as answer]
The behaviour you describe is correct! ofc it cannot deliver you offsets it
doesn't have. when it truncates the data it leaves the newest data intact
so the latest offset is always retained. I don't see what the actual
problem is.

On Mon, 17 Jul 2023 at 11:52, Tarek Haoula ***@***.***> wrote:

> Hi Karl,
>
> I have attached a screenshot that I hope will make a little more sense.
>
> Here we have a stream with only 6 items.
>
> We also have a single consumer which has requested x-stream-offset of 101
> (artificially increased here as i don't know how to force a stream to trim
> its messages to simulate this situation)
>
> Our consumers, retrieve the last offset from the database and set it as an
> argument x-stream-offset when connecting to the stream.
>
> The consumer will not receive a message until the stream reaches that
> offset.
>
> [image: image]
> <https://user-images.githubusercontent.com/349133/253916856-34bd7e45-f2f5-496c-8863-6f728e1eda41.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/8886#discussioncomment-6466443>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFAV6SEVI5UTYDNIY5DXQUKPBANCNFSM6AAAAAA2MXGJMY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 
*Karl Nilsson*

**Reply by thaoula** [reply]
Hi Karl,

Is there a command to truncate the stream on demand? I could not find anything in the docs.

The reason I ask, is that we are seeing the stream truncating to zero messages.  I am not 100% sure why it is doing this or how to trigger it. I thought it was due to the Max Age = 1D.

When the stream resets to 0 message, our consumers stop receiving messages because their stored offset is greater than the message count of the stream. I am also finding that message count of the stream == the x-stream-offset received.

Finally, If I am understanding your correctly. Let's say the stream had 1000 messages and was truncated to 100 messages. The next change should be at x-stream-offset of 1001. Because currently, we see the next offset as being 101?

Appreciate you taking the time to respond.

Regards,
Tarek

**Reply by acogoluegnes** [reply]
The offset in the same stream should always increase, even if retention kicks in. Please use the `rabbitmq-streams stream_status` command on the stream and provide the output here. This should tell us about the last offset in the stream.

Let's say you have a stream with messages going from offset 0 to 1000, retention kicks in, stream contains only now messages from 500 to 1000. Note offsets do not change: message at offset 642 is still at offset 642 after the retention cleaning. This is where the analogy with an array has limitations (there's no index 0, 1, 2, etc).

With the same "trimmed" stream, imagine a consumer registers at offset 2000 (so well ahead of the latest offset in the stream), the broker will actually register at the very end of the stream ([offset specification](https://www.rabbitmq.com/streams.html#consuming) `next`), so the consumer will remain idle until new messages are published to the stream (offset 1001, 1002, etc).

**Reply by thaoula** [reply]
Hi @acogoluegnes,

Thanks for your clarification. 

Running the command you mentioned shows the following

<img width="865" alt="image" src="https://github.com/rabbitmq/rabbitmq-server/assets/349133/c6b5698d-8aff-4015-a80b-e2acce234915">

So i started to look at our configuration and I think, I tracked this down to misconfiguration - missing hostname on our docker compose. We pulled rabbitmq 3.12.1 today which resulted in a new container hostname and new database. I have added a fixed hostname to the config now.

Many thanks @kjnilsson and @acogoluegnes for your responses.

Regards,
Tarek

**Reply by acogoluegnes** [reply]
OK, so the stored offset was for a previous "instance" of the stream. That cannot work indeed :-)

**Reply by thaoula** [reply]
Hi @acogoluegnes,

Yes, pulling the new container generate a new hostname because the docker config did not specify a fixed hostname. This would create a new database and empty stream.

When I run the stream_status command you suggested -
1. How would I tell if the stream has been truncated? 
2. Also, when does the truncation actually happen. Is this a scheduled process?

Also, many thanks for your help so far. Your clarifications help track down the issue.

Regards,
Tarek

**Reply by acogoluegnes** [reply]
If the first offset is not 0, it's very likely the stream has been truncated (see `first_offset` column from `stream_status`).

Retention kicks in when the current segment file reaches a given size, is closed, and a new segment file is open. At this point, the broker checks if some of the "old" segment files can be deleted.

**Reply by acogoluegnes** [reply]
There's also a scheduled job that runs every hour to enforce retention (thanks @kjnilsson for reminding me this).
